packages/synapse-react-client/src/utils/hooks/useOneSageURL.ts

```diff
@@ -1,11 +1,70 @@
-import { useSynapseContext } from '../context'
 import { useMemo } from 'react'
+import { useSynapseContext } from '../context'
+import { BackendDestinationEnum, getEndpoint } from '../functions/index'
 import {
   ONE_SAGE_APPID_QUERY_PARAM_KEY,
   ONE_SAGE_PRODUCTION_URL,
   ONE_SAGE_STAGING_URL,
+  SYNAPSE_BACKEND_PRODUCTION_URL,
+  SYNAPSE_BACKEND_STAGING_URL,
 } from '../SynapseConstants'
 
+export function getOneSageBaseUrl(
+  currentHostname: string,
+  currentBackendEndpoint: string,
+) {
+  // If we're on synapse.org, then go to the accounts site that matches the current backend stack.
+  if (
+    currentHostname === 'synapse.org' ||
+    currentHostname.endsWith('.synapse.org')
+  ) {
+    if (currentBackendEndpoint === SYNAPSE_BACKEND_PRODUCTION_URL) {
+      return ONE_SAGE_PRODUCTION_URL
+    }
+    if (currentBackendEndpoint === SYNAPSE_BACKEND_STAGING_URL) {
+      return ONE_SAGE_STAGING_URL
+    }
+  }
```

Comment:
Before, we were checking each SWC domain. This didn't work for non-SWC sites. For example, if OneSage _staging_ discovers at login time that your password is weak, it redirects you to OneSage _production_.

This should prevent stack hopping and work more reliably across all sites, including OneSage & portals.